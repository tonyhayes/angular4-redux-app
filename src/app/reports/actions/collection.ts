import { Action } from '@ngrx/store';
import { Report } from '../models/report';

export const ADD_REPORT = '[Collection] Add Report';
export const ADD_REPORT_SUCCESS = '[Collection] Add Report Success';
export const ADD_REPORT_FAIL = '[Collection] Add Report Fail';
export const REMOVE_REPORT = '[Collection] Remove Report';
export const REMOVE_REPORT_SUCCESS = '[Collection] Remove Report Success';
export const REMOVE_REPORT_FAIL = '[Collection] Remove Report Fail';
export const LOAD = '[Collection] Load';
export const LOAD_SUCCESS = '[Collection] Load Success';
export const LOAD_FAIL = '[Collection] Load Fail';

/**
 * Add Report to Collection Actions
 */
export class AddReport implements Action {
  readonly type = ADD_REPORT;

  constructor(public payload: Report) {}
}

export class AddReportSuccess implements Action {
  readonly type = ADD_REPORT_SUCCESS;

  constructor(public payload: Report) {}
}

export class AddReportFail implements Action {
  readonly type = ADD_REPORT_FAIL;

  constructor(public payload: Report) {}
}

/**
 * Remove Report from Collection Actions
 */
export class RemoveReport implements Action {
  readonly type = REMOVE_REPORT;

  constructor(public payload: Report) {}
}

export class RemoveReportSuccess implements Action {
  readonly type = REMOVE_REPORT_SUCCESS;

  constructor(public payload: Report) {}
}

export class RemoveReportFail implements Action {
  readonly type = REMOVE_REPORT_FAIL;

  constructor(public payload: Report) {}
}

/**
 * Load Collection Actions
 */
export class Load implements Action {
  readonly type = LOAD;
}

export class LoadSuccess implements Action {
  readonly type = LOAD_SUCCESS;

  constructor(public payload: Report[]) {}
}

export class LoadFail implements Action {
  readonly type = LOAD_FAIL;

  constructor(public payload: any) {}
}

export type Actions =
  | AddReport
  | AddReportSuccess
  | AddReportFail
  | RemoveReport
  | RemoveReportSuccess
  | RemoveReportFail
  | Load
  | LoadSuccess
  | LoadFail;
