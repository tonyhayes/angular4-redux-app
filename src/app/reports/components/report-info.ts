import { Component, Input } from '@angular/core';

import { Report } from '../models/report';

@Component({
  selector: 'analytics-report-info',
  template: `
    <h5 mat-subheader>Report Info:</h5>
    <div *ngIf="scheduled">
      Next Run: {{ nextRun | date:'medium' }}
    </div>
    <div>
      Last Run: {{ lastRun | date:'medium' }}
    </div>
    <div>
      Status: {{ status }}
    </div>
  `,
  styles: [
    `
    h5 {
      margin-bottom: 5px;
    }
  `,
  ],
})
export class ReportInfoComponent {
  @Input() report: Report;
  @Input() scheduled: boolean = false;

  get nextRun() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.nextRun;
    }
    return Date.now();
  }
  get lastRun() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.lastRun;
    }
    return Date.now();
  }
  get status() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.status;
    }
    return 'Normal';
  }
}
