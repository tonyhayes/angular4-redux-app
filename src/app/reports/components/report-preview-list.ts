import { Component, Input } from '@angular/core';
import { Report } from '../models/report';

@Component({
  selector: 'analytics-report-preview-list',
  template: `
    <analytics-report-preview *ngFor="let report of reports" [report]="report" [scheduled]="scheduled"></analytics-report-preview>
  `,
  styles: [
    `
    :host {
      display: flex;
      flex-wrap: wrap;
      justify-content: center;
    }
  `,
  ],
})
export class ReportPreviewListComponent {
  @Input() reports: Report[];
  @Input() scheduled: boolean = false;
}
