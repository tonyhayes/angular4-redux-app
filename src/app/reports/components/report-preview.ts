import { Component, Input } from '@angular/core';
import { Report } from '../models/report';

@Component({
  selector: 'analytics-report-preview',
  template: `
    <a [routerLink]="['/reports', id]">
      <mat-card>
        <mat-card-title-group>
          <mat-card-title>{{ name | bcEllipsis:35 }}</mat-card-title>
        </mat-card-title-group>
        <mat-card-content>
          <p>Report Path : {{ sourceFile | bcEllipsis:40 }}</p>
          <p>Output Path : {{ outputLocation | bcEllipsis }}</p>
        </mat-card-content>
        <mat-card-content *ngIf="scheduled">
          <h5 mat-subheader>Schedule Info:</h5>
          <p>Repeats : {{ repeats }}</p>
          <p>End By : {{ repeatsEndBy }}</p>
          <p>Start Time : {{ repeatTimeStart | date:'medium' }}</p>
          <p>End Time : {{ repeatTimeEnds | date:'medium' }}</p>
        </mat-card-content>
        <mat-card-footer>
          <analytics-report-info [report]="report" [scheduled]="scheduled"></analytics-report-info>
        </mat-card-footer>
      </mat-card>
    </a>
  `,
  styles: [
    `
    :host {
      display: flex;
    }

    :host a {
      display: flex;
    }

    mat-card {
      width: 400px;
      margin: 15px;
      display: flex;
      flex-flow: column;
      justify-content: space-between;
    }

    @media only screen and (max-width: 768px) {
      mat-card {
        margin: 15px 0 !important;
      }
    }
    mat-card:hover {
      box-shadow: 3px 3px 16px -2px rgba(0, 0, 0, .5);
    }
    mat-card-title {
      margin-right: 10px;
    }
    mat-card-title-group {
      margin: 0;
    }
    a {
      color: inherit;
      text-decoration: none;
    }
    mat-card-content {
      margin-top: 15px;
      margin: 15px 0 0;
    }
    span {
      display: inline-block;
      font-size: 13px;
    }
    mat-card-footer {
      padding: 0 25px 25px;
    }
    h5 {
      margin-bottom: 5px;
      font-size: 13px;
    }
  `,
  ],
})
export class ReportPreviewComponent {
  @Input() report: Report;
  @Input() scheduled: boolean = false;

  get id() {
    return this.report.id;
  }

  get name() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.name;
    }
    return 'Big Report I Run All The Time';
  }

  get repeats() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.repeats.repeats;
    }
    return 'Daily';
  }
  get repeatsEndBy() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.repeats.repeatsEndBy;
    }
    return 'Never';
  }
  get repeatTimeStart() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.repeats.repeatTimeStart;
    }
    return Date.now().toString();
  }
  get repeatTimeEnds() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.repeats.repeatTimeEnds;
    }
    return Date.now().toString();
  }

  get sourceFile() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.name;
    }
    return '/path/to/report';
  }
  get outputLocation() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.name;
    }
    return '/output/path/to/report';
  }


}
