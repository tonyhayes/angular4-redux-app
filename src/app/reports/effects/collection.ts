import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/toArray';
import { Injectable } from '@angular/core';
import { Action } from '@ngrx/store';
import { Effect, Actions } from '@ngrx/effects';
import { Database } from '@ngrx/db';
import { Observable } from 'rxjs/Observable';
import { defer } from 'rxjs/observable/defer';
import { of } from 'rxjs/observable/of';

import * as collection from '../actions/collection';
import { Report } from '../models/report';

@Injectable()
export class CollectionEffects {
  /**
   * This effect does not yield any actions back to the store. Set
   * `dispatch` to false to hint to @ngrx/effects that it should
   * ignore any elements of this effect stream.
   *
   * The `defer` observable accepts an observable factory function
   * that is called when the observable is subscribed to.
   * Wrapping the database open call in `defer` makes
   * effect easier to test.
   */
  @Effect({ dispatch: false })
  openDB$: Observable<any> = defer(() => {
    return this.db.open('reports_app');
  });

  @Effect()
  loadCollection$: Observable<Action> = this.actions$
    .ofType(collection.LOAD)
    .switchMap(() =>
      this.db
        .query('reports')
        .toArray()
        .map((reports: Report[]) => new collection.LoadSuccess(reports))
        .catch(error => of(new collection.LoadFail(error)))
    );

  @Effect()
  addReportToCollection$: Observable<Action> = this.actions$
    .ofType(collection.ADD_REPORT)
    .map((action: collection.AddReport) => action.payload)
    .mergeMap(report =>
      this.db
        .insert('reports', [report])
        .map(() => new collection.AddReportSuccess(report))
        .catch(() => of(new collection.AddReportFail(report)))
    );

  @Effect()
  removeReportFromCollection$: Observable<Action> = this.actions$
    .ofType(collection.REMOVE_REPORT)
    .map((action: collection.RemoveReport) => action.payload)
    .mergeMap(report =>
      this.db
        .executeWrite('reports', 'delete', [report.id])
        .map(() => new collection.RemoveReportSuccess(report))
        .catch(() => of(new collection.RemoveReportFail(report)))
    );

  constructor(private actions$: Actions, private db: Database) {}
}
