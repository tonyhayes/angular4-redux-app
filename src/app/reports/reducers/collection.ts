import * as collection from '../actions/collection';

export interface State {
  loaded: boolean;
  loading: boolean;
  ids: string[];
}

const initialState: State = {
  loaded: false,
  loading: false,
  ids: [],
};

export function reducer(
  state = initialState,
  action: collection.Actions
): State {
  switch (action.type) {
    case collection.LOAD: {
      return {
        ...state,
        loading: true,
      };
    }

    case collection.LOAD_SUCCESS: {
      return {
        loaded: true,
        loading: false,
        ids: action.payload.map(report => report.id),
      };
    }

    case collection.ADD_REPORT_SUCCESS:
    case collection.REMOVE_REPORT_FAIL: {
      if (state.ids.indexOf(action.payload.id) > -1) {
        return state;
      }

      return {
        ...state,
        ids: [...state.ids, action.payload.id],
      };
    }

    case collection.REMOVE_REPORT_SUCCESS:
    case collection.ADD_REPORT_FAIL: {
      return {
        ...state,
        ids: state.ids.filter(id => id !== action.payload.id),
      };
    }

    default: {
      return state;
    }
  }
}

export const getLoaded = (state: State) => state.loaded;

export const getLoading = (state: State) => state.loading;

export const getIds = (state: State) => state.ids;
