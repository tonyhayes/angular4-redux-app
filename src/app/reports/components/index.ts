import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';

import { ReportInfoComponent } from './report-info';
import { ReportDetailComponent } from './report-detail';
import { ReportPreviewComponent } from './report-preview';
import { ReportPreviewListComponent } from './report-preview-list';
import { ReportSearchComponent } from './report-search';

import { PipesModule } from '../../shared/pipes';
import { MaterialModule } from '../../material';

export const COMPONENTS = [
  ReportInfoComponent,
  ReportDetailComponent,
  ReportPreviewComponent,
  ReportPreviewListComponent,
  ReportSearchComponent,
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    RouterModule,
    PipesModule,
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class ComponentsModule {}
