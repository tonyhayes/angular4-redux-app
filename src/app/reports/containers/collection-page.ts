import 'rxjs/add/operator/let';
import { Component, ChangeDetectionStrategy, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';

import * as fromReports from '../reducers';
import * as collection from '../actions/collection';
import { Report } from '../models/report';

@Component({
  selector: 'analytics-collection-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <mat-card>
      <mat-card-title>My Scheduled Reports</mat-card-title>
    </mat-card>

    <analytics-report-preview-list [reports]="reports$ | async" [scheduled]=true></analytics-report-preview-list>
  `,
  /**
   * Container components are permitted to have just enough styles
   * to bring the view together. If the number of styles grow,
   * consider breaking them out into presentational
   * components.
   */
  styles: [
    `
    mat-card-title {
      display: flex;
      justify-content: center;
    }
  `,
  ],
})
export class CollectionPageComponent implements OnInit {
  reports$: Observable<Report[]>;

  constructor(private store: Store<fromReports.State>) {
    this.reports$ = store.select(fromReports.getReportCollection);
  }

  ngOnInit() {
    this.store.dispatch(new collection.Load());
  }
}
