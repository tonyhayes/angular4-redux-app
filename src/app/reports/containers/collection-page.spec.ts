import { CollectionPageComponent } from './collection-page';
import { combineReducers, Store, StoreModule } from '@ngrx/store';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { MatCardModule, MatInputModule } from '@angular/material';
import { ReportPreviewListComponent } from '../components/report-preview-list';
import { ReportPreviewComponent } from '../components/report-preview';
import * as collection from '../actions/collection';
import * as fromReports from '../reducers';
import { EllipsisPipe } from '../../shared/pipes/ellipsis';
import { AddCommasPipe } from '../../shared/pipes/add-commas';
import { ReportInfoComponent } from '../components/report-info';

describe('Collection Page', () => {
  let fixture: ComponentFixture<CollectionPageComponent>;
  let store: Store<fromReports.State>;
  let instance: CollectionPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        NoopAnimationsModule,
        StoreModule.forRoot({
          reports: combineReducers(fromReports.reducers),
        }),
        MatCardModule,
        MatInputModule,
        RouterTestingModule,
      ],
      declarations: [
        CollectionPageComponent,
        ReportPreviewListComponent,
        ReportPreviewComponent,
        ReportInfoComponent,
        AddCommasPipe,
        EllipsisPipe,
      ],
    });

    fixture = TestBed.createComponent(CollectionPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.get(Store);

    spyOn(store, 'dispatch').and.callThrough();
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a collection.Load on init', () => {
    const action = new collection.Load();

    fixture.detectChanges();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
