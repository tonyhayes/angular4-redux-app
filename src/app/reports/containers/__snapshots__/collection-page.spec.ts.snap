// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Page should compile 1`] = `
<analytics-collection-page
  reports$={[Function Store]}
  store={[Function Store]}
>
  
      <mat-card
    _ngcontent-c0=""
    class="mat-card"
  >
    
        
    <mat-card-title
      _ngcontent-c0=""
      class="mat-card-title"
    >
      My Scheduled Reports
    </mat-card-title>
    
      
  </mat-card>
  
      <analytics-report-preview-list
    _ngcontent-c0=""
    _nghost-c2=""
    ng-reflect-reports=""
    ng-reflect-scheduled="true"
  >
    
      
    
    
    
  </analytics-report-preview-list>
    
</analytics-collection-page>
`;
